papers/plutus-platform/ledger.tex

```diff
@@ -0,0 +1,238 @@
+This would not be possible if we followed Ethereum's actor model and had contract instances with global state.
+\end{requirement}
+
+\begin{requirement}[Determinism]
+\label{req:ledger-determinism}
+We would like our ledger rules to be \emph{deterministic}, that is, the outcome of transaction validation is entirely determined by the transaction in question (and the current state of the ledger of course).
+This means that the whole process can be simulated accurately by the user \emph{before} submitting a transaction: thus both the outcome of validation and the amount of resources consumed can be determined ahead of time.
+This is helpful for systems that charge for \gls{script} execution, since users can reliably compute how much they will need to pay ahead of time.
+
+A common way for systems to violate this property is by providing \glspl{script} with access to some piece of mutable information, such as the current time (in our system, the current slot has this role).
+\Glspl{script} can then branch on this information, leading to non-deterministic behaviour.
+\end{requirement}
+
+\begin{requirement}[Inspectability]
+\label{req:ledger-inspectability}
+As far as possible, we would like the data which we put on the ledger to be \emph{inspectable}.
+
+This is primarily important for \gls{off-chain}.
+For example, when implementing \glspl{app} which are state machines, we will use the \glspl{datum} of outputs to store the on-chain \emph{state} of the machine.
+If the \gls{off-chain} wants to submit at transaction to advance such a state machine, it may need to read the current state from the chain (after all, the last transition might not have been performed by this agent).
+Hence the \gls{off-chain} needs to be able to inspect the \gls{datum} and interpret it as a value it can work with.
+\end{requirement}
+
+\begin{requirement}[UTXO size]
+\label{req:ledger-utxo-size}
+\Glspl{slot-leader} are likely to be faced with resource constraints when doing transaction validation.
+A key source of memory usage is that the whole set of current \gls{utxo} entries must be kept in memory at all times to check for double-spending.
+Hence, it is important to keep the size of \gls{utxo} entries small.
+\end{requirement}
+
+\subsection{The \glsentryname{data} type}
+\label{sec:data}
+
+We will later require the ability to pass arguments to \glspl{script}.
+However, this requires us to turn such arguments into values in our scripting language of the appropriate type.
+
+In our original design, this was up to the user.
+The \gls{redeemer} and \gls{datum} were themselves scripts, and so it was the user's job to encode their arguments into scripts appropriately.
+
+This did actually work reasonably well (the lifting machinery in \cref{sec:plutus-tx-lifting} was used to great effect), but it was very difficult to go in the opposite direction and turn the scripts back into Haskell values.
+This led us to formulate \cref{req:ledger-inspectability}.
```

Comment:
I do think that part of the goal is to answer "why is it this way?". And in this case I think the historical context is useful for that: it is this way because we tried a perhaps more obvious alternative and found that it didn't work. But maybe that's clear enough by saying we need inspectability...

Reply:
Okay, I think I agree with you.